components/nimbus/src/persistence.rs

```diff
@@ -257,6 +312,104 @@ impl Database {
         self.meta_store
             .put(&mut writer, DB_KEY_DB_VERSION, &DB_VERSION)?;
         writer.commit()?;
+        log::debug!("transaction commited");
+        Ok(())
+    }
+
+    /// Migrates a v1 database to v2
+    ///
+    /// Note that any Err returns (including stuff propagated up via the ?
+    /// operator) will cause maybe_update to assume that this is unrecoverable
+    /// and wipe the database, removing people from any existing enrollments
+    /// and blowing away their experiment history.
+    fn migrate_v1_to_v2(&self, mut writer: &mut Writer) -> Result<()> {
+        log::info!("Upgrading from version 1 to version 2");
+
+        // use try_collect_all to read everything except records that serde
+        // returns deserialization errors on.  Some logging of those errors
+        // happens, but it's not ideal.
+        let reader = self.read()?;
+        let enrollments: Vec<ExperimentEnrollment> =
+            self.enrollment_store.try_collect_all(&reader)?;
+        let experiments: Vec<Experiment> = self.experiment_store.try_collect_all(&reader)?;
+
+        // figure out which enrollments have records that need to be dropped
+        // and log that we're going to drop them and why
+        let slugs_without_enrollment_feature_ids: HashSet<String> = enrollments
+            .iter()
+            .filter_map(
+                    |e| {
+                if matches!(e.status, EnrollmentStatus::Enrolled {ref feature_id, ..} if feature_id.is_empty()) {
+                    log::warn!("Enrollment for {:?} missing feature_ids; experiment & enrollment will be discarded", &e.slug);
+                    Some(e.slug.to_owned())
+                } else {
+                    None
+                }
```

Comment:
I am nervous about this for future migrations. Thinking aloud:

1. `ExperimentEnrollment` is an internal data structure, so we can change it if we want.
2. Enrollments are becoming increasingly detached from feature id (our recent trip into `evolve_enrollments` showed us that features and experiments are deliberately orthogonal.
3. We will want to support multiple `feature_ids`, so it may be an idea to remove `feature_id` and replace it with`feature_ids` in enrollments, or remove it from enrollments altogether.
4. This code will not work as a transitory state between from 1 to (a currently futuristic) 3.

I'm not sure there's anything to do right now, but we should be aware of this if/when we remove `feature_id` from enrollment.

Reply:
Good point.